pkg/storage/stores/tsdb/head_manager.go

```diff
@@ -134,7 +143,78 @@ func NewHeadManager(logger log.Logger, dir string, metrics *Metrics, tsdbManager
 	return m
 }
 
+func (m *HeadManager) loop() {
+	defer m.wg.Done()
+
+	buildPrev := func() error {
+		if err := m.buildTSDBFromWAL(m.prev.initialized); err != nil {
+			return errors.Wrap(err, "building tsdb head")
+		}
+
+		// Now that the tsdbManager has the updated TSDBs, we can remove our references
+		m.mtx.Lock()
+		defer m.mtx.Unlock()
+		m.prevHeads = nil
+		m.prev = nil
+
+		return nil
+	}
+
+	ticker := time.NewTicker(defaultRotationCheckPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			// retry tsdb build failures from previous run
+			if m.prev != nil {
+				err := buildPrev()
+				if err != nil {
+					level.Error(m.log).Log(
+						"msg", "failed building tsdb head",
+						"period", m.period.PeriodFor(m.prev.initialized),
+						"err", err,
+					)
+					// rotating head without building prev would result in loss of index for that period (until restart)
+					continue
+				}
+			}
+
+			now := time.Now()
+			if m.period.PeriodFor(now) > m.period.PeriodFor(m.activeHeads.start) {
+				if err := m.Rotate(now); err != nil {
+					level.Error(m.log).Log(
+						"msg", "failed rotating tsdb head",
+						"err", err,
+					)
+					m.metrics.tsdbHeadRotationsFailedTotal.Inc()
+					continue
+				}
+
+				m.metrics.tsdbHeadRotationsTotal.Inc()
```

Comment:
This metric isn't being incremented if there's an error, which would be a mistake. Taking @sandeepsukhani 's suggestion on metric refactoring would fix this problem.